File: MapsSDKExamplesObjc/Search/SearchAlongTheRouteViewController.m
/**
 * Copyright (c) 2019 TomTom N.V. All rights reserved.
 *
 * This software is the proprietary copyright of TomTom N.V. and its
 * subsidiaries and may be used for internal evaluation purposes or commercial
 * use strictly subject to separate licensee agreement between you and TomTom.
 * If you are the licensee, you are only permitted to use this Software in
 * accordance with the terms of your license agreement. If you are not the
 * licensee then you are not authorised to use this software in any manner and
 * should immediately return it to TomTom N.V.
 */

#import "SearchAlongTheRouteViewController.h"
#import <TomTomOnlineSDKSearch/TomTomOnlineSDKSearch.h>
#import <TomTomOnlineSDKRouting/TomTomOnlineSDKRouting.h>
#import <TomTomOnlineSDKMaps/TomTomOnlineSDKMaps.h>

@interface SearchAlongTheRouteViewController () <TTRouteResponseDelegate,
                                                 TTAlongRouteSearchDelegate>
@property(nonatomic, strong) TTRoute *routePlanner;
@property(nonatomic, strong) TTAlongRouteSearch *alongRouteSearch;
@property(nonatomic, strong) TTMapRoute *mapRoute;
@end

@implementation SearchAlongTheRouteViewController

- (void)setupInitialCameraPosition {
  if (self.mapView.routeManager.routes.count == 0) {
    [super setupInitialCameraPosition];
  } else {
    [self displayRouteOverview];
  }
}

- (OptionsView *)getOptionsView {
  return [[OptionsViewSingleSelect alloc]
      initWithLabels:@[ @"Car repair", @"Gas stations", @"EV stations" ]
          selectedID:-1];
}

- (void)onMapReady {
  [super onMapReady];
  TTRouteQuery *query =
      [[TTRouteQueryBuilder createWithDest:[TTCoordinate HAARLEM]
                                   andOrig:[TTCoordinate AMSTERDAM]] build];
  [self.routePlanner planRouteWithQuery:query];
  [self.progress show];
}

- (void)viewDidLoad {
  [super viewDidLoad];
  self.routePlanner = [TTRoute new];
  self.alongRouteSearch = [TTAlongRouteSearch new];
  self.routePlanner.delegate = self;
  self.alongRouteSearch.delegate = self;
}

#pragma mark OptionsViewDelegate

- (void)displayExampleWithID:(NSInteger)ID on:(BOOL)on {
  [super displayExampleWithID:ID on:on];
  [self.progress show];
  switch (ID) {
  case 2:
    [self searchForEVStations];
    break;
  case 1:
    [self searchForGasStations];
    break;
  default:
    [self searchForCarRepair];
    break;
  }
}

#pragma mark Example

- (void)searchForCarRepair {
  TTAlongRouteSearchQuery *query =
      [[[TTAlongRouteSearchQueryBuilder withTerm:@"REPAIR_FACILITY"
                                       withRoute:self.mapRoute
                               withMaxDetourTime:900] withLimit:10] build];
  [self.alongRouteSearch searchWithQuery:query];
}

- (void)searchForGasStations {
  TTAlongRouteSearchQuery *query =
      [[[TTAlongRouteSearchQueryBuilder withTerm:@"REPAIR_FACILITY"
                                       withRoute:self.mapRoute
                               withMaxDetourTime:900] withLimit:10] build];
  [self.alongRouteSearch searchWithQuery:query];
}

- (void)searchForEVStations {
  TTAlongRouteSearchQuery *query =
      [[[TTAlongRouteSearchQueryBuilder withTerm:@"ELECTRIC_VEHICLE_STATION"
                                       withRoute:self.mapRoute
                               withMaxDetourTime:900] withLimit:10] build];
  [self.alongRouteSearch searchWithQuery:query];
}

#pragma mark TTRouteResponseDelegate

- (void)route:(TTRoute *)route completedWithResult:(TTRouteResult *)result {
  if (!result.routes.firstObject) {
    return;
  }
  self.mapRoute =
      [TTMapRoute routeWithCoordinatesData:result.routes.firstObject
                            withRouteStyle:TTMapRouteStyle.defaultActiveStyle
                                imageStart:TTMapRoute.defaultImageDeparture
                                  imageEnd:TTMapRoute.defaultImageDestination];
  [self.mapView.routeManager addRoute:self.mapRoute];
  [self.mapView.routeManager bringToFrontRoute:self.mapRoute];
  [self displayRouteOverview];
  [self.progress hide];
}

- (void)route:(TTRoute *)route
    completedWithResponseError:(TTResponseError *)responseError {
  [self handleError:responseError];
}

#pragma mark TTAlongRouteSearchDelegate

- (void)search:(TTAlongRouteSearch *)search
    completedWithResponse:(TTAlongRouteSearchResponse *)response {
  [self.progress hide];
  [self.mapView.annotationManager removeAllAnnotations];
  for (TTAlongRouteSearchResult *result in response.results) {
    TTAnnotation *annotation =
        [TTAnnotation annotationWithCoordinate:result.position];
    annotation.selectable = YES;
    [self.mapView.annotationManager addAnnotation:annotation];
  }
}

- (void)search:(TTAlongRouteSearch *)search
    failedWithError:(TTResponseError *)error {
  [self handleError:error];
}

@end

File: MapsSDKExamplesSwift/Search/SearchAlongTheRouteViewController.swift
/**
 * Copyright (c) 2019 TomTom N.V. All rights reserved.
 *
 * This software is the proprietary copyright of TomTom N.V. and its subsidiaries and may be used
 * for internal evaluation purposes or commercial use strictly subject to separate licensee
 * agreement between you and TomTom. If you are the licensee, you are only permitted to use
 * this Software in accordance with the terms of your license agreement. If you are not the
 * licensee then you are not authorised to use this software in any manner and should
 * immediately return it to TomTom N.V.
 */

import MapsSDKExamplesCommon
import MapsSDKExamplesVC
import TomTomOnlineSDKMaps
import TomTomOnlineSDKRouting
import TomTomOnlineSDKSearch
import UIKit

class SearchAlongTheRouteViewController: RoutingBaseViewController, TTRouteResponseDelegate, TTAlongRouteSearchDelegate {
    let routePlanner = TTRoute()
    let alongRouteSearch = TTAlongRouteSearch()
    var mapRoute: TTMapRoute!

    override func setupCenterOnWillHappen() {
        if mapView.routeManager.routes.isEmpty {
            super.setupCenterOnWillHappen()
        } else {
            displayRouteOverview()
        }
    }

    override func getOptionsView() -> OptionsView {
        return OptionsViewSingleSelect(labels: ["Car repair", "Gas stations", "EV stations"], selectedID: -1)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        routePlanner.delegate = self
        alongRouteSearch.delegate = self

        let query = TTRouteQueryBuilder.create(withDest: TTCoordinate.HAARLEM(), andOrig: TTCoordinate.AMSTERDAM())
            .build()
        routePlanner.plan(with: query)
        progress.show()
    }

    // MARK: OptionsViewDelegate

    override func displayExample(withID ID: Int, on: Bool) {
        super.displayExample(withID: ID, on: on)
        progress.show()
        switch ID {
        case 2:
            searchForEVStations()
        case 1:
            searchForGasStations()
        default:
            searchForCarRepair()
        }
    }

    // MARK: Examples

    func searchForCarRepair() {
        let query = TTAlongRouteSearchQueryBuilder(term: "REPAIR_FACILITY", withRoute: mapRoute, withMaxDetourTime: 900)
            .withLimit(10)
            .build()
        alongRouteSearch.search(with: query)
    }

    func searchForGasStations() {
        let query = TTAlongRouteSearchQueryBuilder(term: "PETROL_STATION", withRoute: mapRoute, withMaxDetourTime: 900)
            .withLimit(10)
            .build()
        alongRouteSearch.search(with: query)
    }

    func searchForEVStations() {
        let query = TTAlongRouteSearchQueryBuilder(term: "ELECTRIC_VEHICLE_STATION", withRoute: mapRoute, withMaxDetourTime: 900)
            .withLimit(10)
            .build()
        alongRouteSearch.search(with: query)
    }

    // MARK: TTRouteResponseDelegate

    func route(_: TTRoute, completedWith result: TTRouteResult) {
        guard let plannedRoute = result.routes.first else {
            return
        }
        mapRoute = TTMapRoute(coordinatesData: plannedRoute,
                              with: TTMapRouteStyle.defaultActive(),
                              imageStart: TTMapRoute.defaultImageDeparture(),
                              imageEnd: TTMapRoute.defaultImageDestination())
        mapView.routeManager.add(mapRoute)
        mapView.routeManager.bring(toFrontRoute: mapRoute)
        displayRouteOverview()
        progress.hide()
    }

    func route(_: TTRoute, completedWith responseError: TTResponseError) {
        handleError(responseError)
    }

    // MARK: TTAlongRouteSearchDelegate

    func search(_: TTAlongRouteSearch, completedWith response: TTAlongRouteSearchResponse) {
        progress.hide()
        mapView.annotationManager.removeAllAnnotations()
        for result in response.results {
            let annotation = TTAnnotation(coordinate: result.position)
            annotation.selectable = false
            mapView.annotationManager.add(annotation)
        }
    }

    func search(_: TTAlongRouteSearch, failedWithError error: TTResponseError) {
        handleError(error)
    }
}
